test/data.jl

```diff
@@ -473,4 +477,109 @@ end
     @test isapprox(df, DataFrame(x1=[0.0, 0.0, 0.0], x2=[1.1, 1.0, 1.0]), atol=0.11)
 end
 
+@testset "completecombinations"  begin
+    df1 = DataFrame(a=[1, 2, missing], b=[1, 1, 2],
+                    c=categorical([11, 12, missing]), d=111:113)
+    levels!(df1.c, [12, 11, 10])
+
+    for ad in (true, false)
+        @test isequal_coltyped(completecombinations(df1, [:a, :b], allowduplicates=ad),
+                               DataFrame(a=[1, 2, missing, 1, 2, missing],
+                                         b=[1, 1, 1, 2, 2, 2],
+                                         c=categorical([11, 12, missing, missing, missing, missing]),
+                                         d=[111, 112, missing, missing, missing, 113]))
+        @test isequal_coltyped(completecombinations(df1, [:a, :b], fill="X", allowduplicates=ad),
+                               DataFrame(a=[1, 2, missing, 1, 2, missing],
+                                         b=[1, 1, 1, 2, 2, 2],
+                                         c=[11, 12, "X", "X", "X", missing],
+                                         d=[111, 112, "X", "X", "X", 113]))
+        @test isequal_coltyped(completecombinations(df1, :c, allowduplicates=ad),
+                               DataFrame(a=[2, 1, missing, missing],
+                                         b=[1, 1, missing, 2],
+                                         c=categorical([12, 11, 10, missing]),
+                                         d=[112, 111, missing, 113]))
+        @test isequal_coltyped(completecombinations(df1, :c, fill="X", allowduplicates=ad),
+                               DataFrame(a=[2, 1, "X", missing],
+                                         b=[1, 1, "X", 2],
+                                         c=categorical([12, 11, 10, missing]),
+                                         d=[112, 111, "X", 113]))
+        @test isequal_coltyped(completecombinations(df1, [:c, :b], allowduplicates=ad),
+                               DataFrame(a=[2; 1; fill(missing, 6)],
+                                         b=[1, 1, 1, 1, 2, 2, 2, 2],
+                                         c=categorical([12, 11, 10, missing, 12, 11, 10, missing]),
+                                         d=[112; 111; fill(missing, 5); 113]))
```

Comment:
I don't see the test for `levels`. Am I missing it?

Reply:
I have added the tests - as usual tricky cases are present.